src/audio/eq_fir.c

```diff
@@ -484,21 +484,37 @@ static int fir_cmd_get_data(struct comp_dev *dev,
 			    struct sof_ipc_ctrl_data *cdata, int max_size)
 {
 	struct comp_data *cd = comp_get_drvdata(dev);
-
+	unsigned char *dst, *src;
+	size_t offset;
 	size_t bs;
 	int ret = 0;
 
 	switch (cdata->cmd) {
 	case SOF_CTRL_CMD_BINARY:
 		trace_eq("fir_cmd_get_data(), SOF_CTRL_CMD_BINARY");
 
+		max_size -= sizeof(struct sof_ipc_ctrl_data) +
+			sizeof(struct sof_abi_hdr);
+
 		/* Copy back to user space */
 		if (cd->config) {
+			src = (unsigned char *)cd->config;
+			dst = (unsigned char *)cdata->data->data;
 			bs = cd->config->size;
-			if (bs > SOF_EQ_FIR_MAX_SIZE || bs == 0 ||
-			    bs > max_size)
-				return -EINVAL;
-			memcpy(cdata->data->data, cd->config, bs);
+			cdata->elems_remaining = 0;
+			offset = 0;
+			if (bs > max_size) {
+				bs = (cdata->msg_index + 1) * max_size > bs ?
+					bs - cdata->msg_index * max_size :
+				  max_size;
+				offset = cdata->msg_index * max_size;
+				cdata->elems_remaining = cd->config->size -
+					offset;
+			}
+			cdata->num_elems = bs;
+			trace_eq("fir_cmd_get_data(), %zu %u %u %zu",
+				 bs, cdata->msg_index, max_size, offset);
+			memcpy(dst, src + offset, bs);
```

Comment:
We should start using memcpy_s now or wait for pr #1103 ?

Reply:
whatever suits better...